### PR TITLE
✨ Ability system 

### LINKED ISSUE #191: ♻️ Overhaul Ability System
Before anyone works on this, please submit a class diagram, and propsed system for implementation.

### PR DESCRIPTION
**Pull request by Timsel1**
## Content

An overhaul of the ability system, it now works based on scriptable objects. Note projectiles are different form all other ability types and need a prefab to work.

### Added
`Assets/Animations/Player/AbilityLaser.anim` : Was created. Temp animation for the laser ability.
`Assets/Scripts/Ability/Ability.cs` : Was created. Ability base class.
`Assets/Scripts/Ability/AbilityManager.cs` : Was created. Manages new ability system.
`Assets/Scripts/Ability/AbilitySlot.cs` : Was created. Enum to check what slot an ability is in.
`Assets/Scripts/Ability/CastTarget.cs` : Was created. Enum to check the cast target of an ability.
`Assets/Scripts/Ability/Combat/DamageAbility.cs` : Was created. Base Damage ability script.
`Assets/Scripts/Ability/Combat/LaserBeam.cs` : Was created. Laser ability logic.
`Assets/Scripts/Ability/Combat/LineAreaAbility.cs` : Was created. LineAreaAbility base class.
`Assets/Scripts/Ability/Combat/ProjectileAbility.cs` : Was created. ProjectileAbility base class.
`Assets/Scripts/Ability/Combat/ProjectileCollisionDamage.cs` : Was created. Handles collision checks and damage for projectile abilities (should be added to projectile prefab).
`Assets/Scripts/Ability/Combat/RogueApeStudio.Crusader.Ability.Combat.asmdef` : Was created. asmdf for combat abilities.
`Assets/Scripts/Ability/Combat/ScriptableObjects.meta` : Was created. Folder for ability scriptable objects.
`Assets/Scripts/Ability/Combat/SmiteAbility.cs` : Was created. Logic for smite ability.
`Assets/Scripts/Ability/Combat/SpearAbility.cs` : Was created. Logic for spear ability.
`Assets/Scripts/Ability/Combat/SphereAbility.cs` : Was created. SphereAbility base class.
`Assets/Scripts/Ability/Combat/WaveAbility.cs` : Was created. Wave ability logic.
`Assets/Scripts/Ability/CursorPositionCast.cs` : Was created. Handles directions for rotations.
`Assets/Scripts/Ability/RogueApeStudio.Crusader.Ability.asmdef` : Was created. asmdf for abilities.

### Updated
`Assets/Animations/Player/AbilitySmite.anim` : Was updated. Updated animation events.
`Assets/Animations/Player/AbilitySpear.anim` : Was updated. Updated animation events.
`Assets/Animations/Player/AbilityWave.anim` : Was updated. Updated animation events.
`Assets/Animations/Player/DiveForward.anim` : Was updated. Removed old animation events.
`Assets/Animations/Player/Player.controller` : Was updated. Added a state for laser animation.
`Assets/Prefabs/Player/Player_1.prefab` : Was updated. Added new ability system.
`Assets/Prefabs/VFX/Abilities/HolySpear.prefab` : Was updated / renamed. Renamed and added new ability system.
`Assets/Scripts/Player/Abilities/WaveAbility.cs` : Was updated. Changed used namespace from units.knockback to just units.
`Assets/Scripts/Player/Combat/AttackDamage.cs` : Was updated. Changed used namespace from units.knockback to just units.
`Assets/Scripts/Units/Knockback.cs` : Was updated. Changed used namespace from units.knockback to just units.

## Testing

The feature / Bugfix can be testing by following these steps:

1. pull branch
2. go to scripts/ability/combat/scriptableObject
3. change stats and see the effect.
4. change current ability by dragging in a different scriptable object on the player prefab in the ability manager script under abilities.

Closes #191

### CODE REVIEW

File: Assets/Scripts/Ability/Ability.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            throw new System.NotImplementedException(nameof(Cast));
+        }
+        
+        public virtual void Cast(Vector3 position, Vector3 direction, Quaternion rotation)
+        {
+            throw new System.NotImplementedException(nameof(Cast));
+        }
+
+        internal async void StartCooldown(CancellationToken token)
```
Comment (NathanThus) [inline Assets/Scripts/Ability/Ability.cs:48]: How are you stopping the cooldown?

File: Assets/Scripts/Ability/AbilityManager.cs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        [Header("Slots")]
+        [SerializeField] private GameObject _slotParent;
+        [SerializeField] private VisualEffect[] _slots;
+
+        private CrusaderInputActions _inputActions;
+        private InputAction _ability1;
+        private InputAction _ability2;
+        private InputAction _ability3;
+        private CancellationTokenSource _cancellationTokenSource;
+        private Dictionary<int, CastTarget> _castTargetCache;
```
Comment (NathanThus) [inline Assets/Scripts/Ability/AbilityManager.cs:37]: Why did you go for a `Dictionary `here? Why not a `2D List`, or a `List<CustomStruct>`?

File: Assets/Scripts/Ability/AbilityManager.cs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                    _playerCharacterTransform.rotation = Quaternion.LookRotation(direction);
+                }
+                else if (castTarget == CastTarget.Position)
+                    _position = _positionCast.GetCursorPosition(_cam, _layerMask);
+            }
+        }
+
+        private bool IsValidCastTarget(CastTarget castTarget)
+        {
+            return castTarget == CastTarget.Self ||
```
Comment (NathanThus) [inline Assets/Scripts/Ability/AbilityManager.cs]: I feel like this could be done better.

File: Assets/Scripts/Ability/AbilityManager.cs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                    throw new NotImplementedException($"{nameof(CastTarget)} enum value is not implemented!");
+            }
+        }
+
+        internal void RemoveEffect()
+        {
+            _slots[_slotNumber].transform.SetParent(null, true);
+        }
+
+        public void Setup(List<Ability> abilities)
```
Comment (NathanThus) [inline Assets/Scripts/Ability/AbilityManager.cs:191]: Please verify that the list is not `null` and `List.count` is not 0.

File: Assets/Scripts/Ability/Combat/LaserBeam.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        private void OnEnable()
+        {
+            _cancellationTokenSource = new();
+        }
+
+        private void OnDisable()
+        {
+            _cancellationTokenSource.Cancel();
+            _cancellationTokenSource?.Dispose();
```
Comment (NathanThus) [inline Assets/Scripts/Ability/Combat/LaserBeam.cs:26]: This **WILL** always be instaniated, or you've done something awful.
Just give this comment a thumbs up for read.

File: Assets/Scripts/Ability/Combat/SmiteAbility.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            foreach (var hitCollider in hitColliders)
+            {
+                if (hitCollider.CompareTag("Enemy"))
+                {
+                    Enemy enemyComponent = hitCollider.GetComponent<Enemy>();
+                    EnemyHealth healthComponent = hitCollider.GetComponent<EnemyHealth>();
+
+                    if (enemyComponent == null)
+                    {
+                        Debug.LogWarning("Enemy component is null, skipping.");
```
Comment (NathanThus) [inline Assets/Scripts/Ability/Combat/SmiteAbility.cs]: Why is it not throwing an `ArgumentNullException`?

**Review by NathanThus — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Ability/Combat/SmiteAbility.cs (hunk 6 above)
Comment (Timsel1) [reply Assets/Scripts/Ability/Combat/SmiteAbility.cs]: full if statement is gone

File: Assets/Scripts/Ability/Ability.cs (hunk 1 above)
Comment (Timsel1) [reply Assets/Scripts/Ability/Ability.cs:48]: verbally resolved

File: Assets/Scripts/Ability/AbilityManager.cs (hunk 4 above)
Comment (Timsel1) [reply Assets/Scripts/Ability/AbilityManager.cs:191]: done

File: Assets/Scripts/Ability/AbilityManager.cs (hunk 2 above)
Comment (NathanThus) [reply Assets/Scripts/Ability/AbilityManager.cs:37]: This has been resolved verbally.

File: Assets/Scripts/Ability/AbilityManager.cs (hunk 3 above)
Comment (Timsel1) [reply Assets/Scripts/Ability/AbilityManager.cs]: completely removed, since it is impossible to select anything other than a valid cast target

**Review by NathanThus — APPROVED** [review]

**Pull request merged**